Add a chapter about maps to the Efficiency Guide

bjorng opened the pull request.

dgud committed:
ssl: avoid decoding certs several times

The asn.1 decoding can be expansive avoid it by keeping both der and the
decoded cert available.

dgud committed:
ssl: Add a cert record containing both der and decoded

dgud committed:
ssl: refactor some code

Use the cert record

dgud committed:
ssl: Use cert rec in db

Continuation of the refactoring.

dgud committed:
ssl: Remove dead code

IngelaAndin committed:
public_key: Allow verify_fun to alter expire reason

This allows applications to distinguish between a trusted anchor cert
expiration and a normal cert expiration

IngelaAndin committed:
ssl: Extend certificate chain recreation logic

To be able to verify the certificate chain for more variants of
configuration, we now ignore expired trusted anchors at first.
This so that we can look for other possible valid issuers in a specific path,
that is a cross signed root certificate or another trusted certificate
present in the trust store but shadowed by the one that has expired.
If no alternative is found the connection will fail with a TLS certificate
expire ALERT.

This commit in addition to modifing path validation handling eliminates
unessessary code duplication from tls_handshake_1_3.

Closes #4877

lagebr committed:
doc: Fix spelling errors

legoscia committed:
observer: add button to trace all processes and ports

bmk committed:
[kernel|gen-udp-socket] Add socket based gen_udp

First stage of the 'socket' based gen_udp.

OTP-17410

peterdmv committed:
ssl: Fix signature algorithm selection for ECDSA

The signature selection algorithm has been changed to also verify
if the client supports signatures using the elliptic curve of the
server's public/private key pair.

The old algorithm only verified the signature algorithm and the
hash algorithm, not taking into account the elliptic curve of
the key in the server's certificate. This could result in
successful TLS connections with non-standard algorithm combinations
in the CertificateVerify message such as ECDSA with secp256r1 and
SHA512, when the server's certificate had a public key on an
unsupported elliptic curve, signed with a non-ECDSA signature
algorithm and hash algorithm combination e.g. RSA with SHA256.

peterdmv committed:
ssl: Fix signature algorithm selection for ECDSA

The signature selection algorithm has been changed to also verify
if the client supports signatures using the elliptic curve of the
server's public/private key pair.

The old algorithm only verified the signature algorithm and the
hash algorithm, not taking into account the elliptic curve of
the key in the server's certificate. This could result in
successful TLS connections with non-standard algorithm combinations
in the CertificateVerify message such as ECDSA with secp256r1 and
SHA512, when the server's certificate had a public key on an
unsupported elliptic curve, signed with a non-ECDSA signature
algorithm and hash algorithm combination e.g. RSA with SHA256.

KennethL committed:
support extensionmark in nested SEQUENCE in parameterized type
GH-4902 OTP-17522

IngelaAndin committed:
ssl: Avoid unnecessary ASN1 decodes

IngelaAndin committed:
ssl: Use new inet:monitor/1 to monitor listen sockets

To be able to fully work with new gen_tcp socket backend.

Also Make sure TLS-1.3 server session ticket handler processes
monitor the listen socket so that it can terminate gracefully.

petrkozorezov committed:
fix ssl:close with nif sockets

bmk committed:
[observer] Update app dependencies

OTP-17346

bmk committed:
[runtime_tools] Update app dependencies

OTP-17346

bjorng committed:
Add a chapter about maps to the Efficiency Guide

josevalim committed:
Do not attempt to start children of running apps

When optional applications were added in 79146711,
ensure_all_started/2 had to take explicit control
over the initialization cycle so available optional
applications were started properly. However, in
doing so, we lost one important optimization, which
was to not attempt to start children of a given
application if said application was already running.

This commit address this by exporting is_running/1
and skipping the work in case the given app is already
running. This change makes it so the time taken to
start a deep dependency tree goes from 58s to 4s.

Maria-12648430 committed:
Add no_return() spec to gen_statem:init/1 note

If no spec is given, dialyzer will complain
that init/1 only terminates with explicit
exception.

u3s commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      <p>Therefore, if maps are used instead of records and it is
+      expected that many instances of the map will be created, it is
+      more efficient from a memory standpoint to avoid using large
+      maps.</p>
```
Maybe could be clarified what is actually suggested, interpretation options I see are:
- you're recommending reducing maps size to fit in small,
- you're recommending staying with records if can't reduce the structure size or maybe actually not ;-)?

Regarding records/tuple memory usage in this scenario, maybe a reference could be added to read more on that topic.
If not described elsewhere, maybe some more words could make it easier to decide if having a map instead of record is worth it for bigger structures.

bjorng replied:
I meant that size of the map should reduced by grouping element into sub maps. I have added a clarification.

I deliberately avoided the subject of maps vs records, especially regarding performance. I am not sure that it can be said that one of them is always faster than the one. Perhaps I should add some text given some guidance. (Use records for everything internal where you can't be bitten by the size change of the record when changing code, and maps for everything else.)

u3s commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      marker="stdlib:maps#filtermap/2">maps:filtermap/2</seemfa> is
+      implemented in Erlang. It will build a new map. See the notes
+      for <c>maps:map/2</c> and <c>maps:filter/2</c> for hints on how
+      to implement a more efficent version.</p>
```
efficient

u3s reviewed (commented):
I've added 2 minor comments inline.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      it will require more memory and keys can no longer be shared
+      with other instances of the map.</p></item>
+
+      <item><p>When creating a new map, always create it with the
```
the the

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      <item><p>Keys are usually variables not known at compile-time.</p></item>
+      <item><p>There can be any number of elements in the map.</p></item>
+      <item><p>Usually, no more than one element is looked up or
+      updated at the time.</p></item>
```
Not sure, but I think "at a time" is better.

bjorng replied:
I will change to "at once" because I have used that wording in other places.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+    with a few possible exceptions:</p>
+
+    <list type="bulleted">
+      <item><p>If it is necessary to frequenquently convert a
```
Maybe run a spell checker...

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+
+    <list type="bulleted">
+      <item><p>If it is necessary to frequenquently convert a
+      dictionary to a sorted listed, or from a sorted list to a
```
"sorted list"

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      could be a better choice.</p></item>
+
+      <item><p>If all keys are non-negative integers, the <seeerl
+      marker="stdlib:array">array</seeerl> module could be a better
```
I think the technical writers prefer "can" before "could", but I haven't bothered to check their document...

bjorng replied:
Yes, that sounds vaguely familiar. I'll change.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+
+    <p>Starting in OTP 24, the <seeerl marker="stdlib:sets">sets</seeerl> module
+    has an option to represent its sets as map. Examples:</p>
+
```
"its" is a bit superfluous, I think.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+    <list type="bulleted">
+      <item><p><seemfa
+      marker="stdlib:ordsets#intersection/2">ordsets:intersection/2</seemfa>
+      could be more efficient than <seemfa
```
"can"/"could" again (will not mention it again)

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      </taglist>
+
+      <p>As an example, let us look at how the map <c>#{a => foo, z => bar}</c>
+      would be represented:</p>
```
The technical writer documents says that "would" is to be avoided.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+        <tcaption>#{a => foo, z => bar}</tcaption>
+      </table>
+
+      <p>Let us update the map: <c>M#{q => baz}</c>. The map will
```
The technical writer document says that "will" should be avoided.

bjorng replied:
I have a lot of "will". I will remove some where I'll think it makes sense.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+    #{a => default, b => default, c => default}.]]></code>
+
+      <p>Defined like this, the key tuple <c>{a,b,c}</c> will be a
+      global literal.  To ensure that the key tuple is shared,
```
The comma is not necessary.

uabboli commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+
+      <p>Whether that is faster than calling <c>maps:get/3</c>
+      multiple times depends on the size of the map and the number of
+      default values that must be handled.</p>
```
Can "that must be handled" be removed?

bjorng replied:
Yes. Will do.

bjorng commented:
Thanks for the comments. I have pushed a fixup commit that addresses most of them.

IngelaAndin committed:
ssl: Correct and clearify downgrade

IngelaAndin committed:
ssl: Handle possible session tickets in TLS-1.3 downgrade state

Closes #5009

KennethL committed:
Change the copyright notice in the footer to always include the current year

KennethL committed:
Merge pull request #5034 from KennethL/kenneth/doc_index_copyright

Change the copyright notice in the footer to always include the current year

kjellwinblad committed:
Merge branch 'jv-recur-is-running-apps' into maint

* jv-recur-is-running-apps:
  Do not attempt to start children of running apps

OTP-17519
PR-5017

u3s committed:
ssl: tls_server_session_ticket unit tests added

- test suite added
- enabled coverage verification

bmk committed:
[kernel] Add 'active' to info

Add 'active' to info for gen_tcp_socket and
the old 'port' socket's.

OTP-17410

bmk committed:
[kernel|gen-udp-socket] All test cases passes

Now all existing test cases passes!
Except the one we skip.

OTP-17410

IngelaAndin committed:
ssl: Correct error return value

IngelaAndin committed:
Merge pull request #5030 from IngelaAndin/ingela/ssl/downgrade/GH-5009/OTP-17517

Ingela/ssl/downgrade/gh 5009/otp 17517

bmk committed:
[kernel|test] Tweaking gethostby test case(s)

dgud committed:
Merge pull request #4962 from legoscia/observer-trace-all

observer: add button to trace all processes and ports

OTP-17520

garazdawi committed:
Change travis to github actions in CONTRIBUTING.md

garazdawi committed:
Merge branch 'lukas/fix-contributing-gh-actions' into maint

* lukas/fix-contributing-gh-actions:
  Change travis to github actions in CONTRIBUTING.md

IngelaAndin committed:
Merge branch 'ingela/ssl/error-handling-fix/OTP-17521' into maint

* ingela/ssl/error-handling-fix/OTP-17521:
  ssl: Correct error return value

IngelaAndin committed:
Merge pull request #5050 from IngelaAndin/ingela/ssl/monitor-socket/OTP-17392

ssl: Use new inet:monitor/1 to monitor listen sockets

IngelaAndin committed:
Merge branch 'ingela/ssl/maint-23/alt-certs/GH-4877/OTP-17975' into maint

IngelaAndin committed:
ssl: Do not run upgrade test when version is not changed

IngelaAndin committed:
Merge branch 'ingela/ssl/maint/OTP-17150/alt-certs/GH-4877/OTP-17475' into maint

IngelaAndin committed:
ssl: Handle empty SNI

michaelklishin committed:
Doc typo: build_flavor was renamed to emu_flavor  (for maint)

Same as #5042 but based on `maint`.

garazdawi committed:
Merge pull request #5051 from michaelklishin/patch-4

Doc typo: build_flavor was renamed to emu_flavor  (for maint)

KennethL committed:
Merge branch 'kenneth/asn1/GH-4902/OTP-17522' into maint

* kenneth/asn1/GH-4902/OTP-17522:
  support extensionmark in nested SEQUENCE in parameterized type GH-4902 OTP-17522

uabboli committed:
stdlib: Fix a concerning io_lib option 'chars_limit'

See also https://github.com/erlang/otp/issues/5053.

bmk committed:
[kernel] Dialyzer fixes

OTP-17410

bmk committed:
[kernel] Fix copyright end date

OTP-17410

bmk committed:
[kernel|gen-udp-socket] Dialyzer fixes

Running the checks locally did not detect this; grumble, grumble, ...

OTP-17410

bmk committed:
[kernel|gen-udp-socket|doc] Add preliminary documentation

Add note about 'inet-backend' and the currently only
difference to "standard" (=port) gen_udp.

OTP-17410

bmk committed:
[kernel|gen-udp-socket|test] Added (socket) monitor test case(s)

Basically "copied" the test cases from gen_tcp_misc_SUITE.

OTP-17410

sigsergv committed:
Fix public_key:pkix_sign for RSASSA-PSS signature

uabboli committed:
Merge branch 'hasse/stdlib/chars_limit_fix/GH-5053/OTP-17525' into maint

* hasse/stdlib/chars_limit_fix/GH-5053/OTP-17525:
  stdlib: Fix a concerning io_lib option 'chars_limit'

u3s committed:
Merge branch 'kuba/ssl/tls13_session_ticket_cov/OTP-17394' into maint

* kuba/ssl/tls13_session_ticket_cov/OTP-17394:
  ssl: tls_server_session_ticket unit tests added

bmk committed:
[kernel] Add new module (gen_udp_socket) to app file

OTP-17410

IngelaAndin committed:
ssl: Enhance cookie error handling

"Catch all" to handle partially corrupt data

IngelaAndin committed:
ssl: Better error handling for rejecting legacy

IngelaAndin committed:
ssl: Fix indentation

IngelaAndin committed:
ssl: Ignore invalid max frag extension

IngelaAndin committed:
ssl: Avoid internal error, be graceful!

bmk committed:
[kernel|gen-udp-socket] Use 'tuple_size' for tuple size testing

Use 'tuple_size' instead of 'size' for testing size(s) of tuples
in function guards.

OTP-17410

IngelaAndin committed:
ssl: Update socket dependency for PR-5050, OTP-17392

IngelaAndin committed:
Merge pull request #4979 from peterdmv/ssl/fix-sign-alg-selection

ssl: Fix signature algorithm selection for ECDSA

OTP-17529

IngelaAndin committed:
Merge branch 'ingela/ssl/ECDSA-signatures/OTP-17529' into maint

IngelaAndin committed:
ssl: Remove use of non recommended function size

IngelaAndin committed:
Merge branch 'ingela/ssl/negative-tests/OTP-17530' into maint

* ingela/ssl/negative-tests/OTP-17530:
  ssl: Avoid internal error, be graceful!
  ssl: Ignore invalid max frag extension
  ssl: Fix indentation
  ssl: Better error handling for rejecting legacy
  ssl: Enhance cookie error handling
  ssl: Handle empty SNI

RaimoNiskanen committed:
Merge pull request #5019 from Maria-12648430/patch-2

Add no_return() spec to gen_statem:init/1 doc example

bmk committed:
Merge pull request #4996 from petrkozorezov/ssl/fix-new-socket-api-close

fix ssl:close with nif sockets

RaimoNiskanen committed:
Merge pull request #4948 from lagebr/patch-13

doc: Fix spelling errors

bmk committed:
[kernel|gen-udp-socket] Fixed bind

Don't ignore the port number just because the address is
"undefined".

OTP-17410

sigsergv committed:
Fixing test public_key_SUITE:rsa_pss_sign_verify()

IngelaAndin committed:
ssl: Update test to use new parameter

IngelaAndin committed:
ssl: Update test to use new parameter

IngelaAndin committed:
Merge branch 'ingela/ssl/maint-24-pem-cache-test' into maint

IngelaAndin committed:
Merge branch 'ingela/ssl/maint-23-pem-cache-test' into maint

IngelaAndin committed:
Merge remote-tracking branch 'ingela/ingela/ssl/monitor-socket-dependency' into maint

* ingela/ingela/ssl/monitor-socket-dependency:
  ssl: Update socket dependency for PR-5050, OTP-17392

IngelaAndin committed:
Merge branch 'ingela/ssl/replace-size-calls' into maint

* ingela/ssl/replace-size-calls:
  ssl: Remove use of non recommended function size

IngelaAndin committed:
ssl: Update runtime dependency

IngelaAndin committed:
Merge branch 'ingela/ssl/runtime-dep' into maint

* ingela/ssl/runtime-dep:
  ssl: Update runtime dependency

Commit:
Merge branch 'rickard/schedulers-online-fix/GH-4809/OTP-17500' into maint-23

* rickard/schedulers-online-fix/GH-4809/OTP-17500:
  Fix erlang:system_flag(schedulers_online, _)

Commit:
Merge branch 'rickard/monotonic-hrtime/23/OTP-17493' into maint-23

* rickard/monotonic-hrtime/23/OTP-17493:
  Update configure scripts
  Fix monotonic hrtime configure test

Commit:
Merge branch 'anders/common_test/netconf/OTP-17506/OTP-17509' into maint-23

* anders/common_test/netconf/OTP-17506/OTP-17509:
  Let a receiver of incoming NETCONF notifications be configured
  Simplify xmlns mechanics
  Remove misleading documentation
  Fix badarg on reception of unexpected NETCONF notification

Commit:
Merge branch 'sverker/23/dist-auto-connect-crash/GH-4964/OTP-17513' into maint-23

* sverker/23/dist-auto-connect-crash/GH-4964/OTP-17513:
  erts: Fix write-after-free bug for DistEntry (AGAIN)

Commit:
Merge branch 'ingela/ssl/maint/OTP-17150/alt-certs/GH-4877/OTP-17475' into maint-23

* ingela/ssl/maint/OTP-17150/alt-certs/GH-4877/OTP-17475:
  ssl: Do not run upgrade test when version is not changed
  ssl: Avoid unnecessary ASN1 decodes
  ssl: Extend certificate chain recreation logic
  public_key: Allow verify_fun to alter expire reason
  ssl: Remove dead code
  ssl: Use cert rec in db
  ssl: refactor some code
  ssl: Add a cert record containing both der and decoded
  ssl: avoid decoding certs several times

Commit:
Merge branch 'kenneth/asn1/GH-4902/OTP-17522' into maint-23

* kenneth/asn1/GH-4902/OTP-17522:
  support extensionmark in nested SEQUENCE in parameterized type GH-4902 OTP-17522

Commit:
Merge branch 'hasse/stdlib/chars_limit_fix/GH-5053/OTP-17525' into maint-23

* hasse/stdlib/chars_limit_fix/GH-5053/OTP-17525:
  stdlib: Fix a concerning io_lib option 'chars_limit'

# Conflicts:
#	lib/stdlib/test/io_SUITE.erl

Commit:
Merge branch 'ingela/ssl/ECDSA-signatures/OTP-17529' into maint-23

* ingela/ssl/ECDSA-signatures/OTP-17529:
  ssl: Fix signature algorithm selection for ECDSA

Commit:
Merge branch 'ingela/ssl/runtime-dep' into maint-23

* ingela/ssl/runtime-dep:
  ssl: Update runtime dependency

Commit:
Merge branch 'ingela/ssl/maint-23-pem-cache-test' into maint-23

* ingela/ssl/maint-23-pem-cache-test:
  ssl: Update test to use new parameter

Commit:
Prepare release

Commit:
Updated OTP version

Commit:
Merge branch 'anders/common_test/netconf/OTP-17506/OTP-17509' into maint-24

* anders/common_test/netconf/OTP-17506/OTP-17509:
  Let a receiver of incoming NETCONF notifications be configured
  Simplify xmlns mechanics
  Remove misleading documentation
  Fix badarg on reception of unexpected NETCONF notification

Commit:
Merge branch 'ingela/ssl/maint/OTP-17150/alt-certs/GH-4877/OTP-17475' into maint-24

* ingela/ssl/maint/OTP-17150/alt-certs/GH-4877/OTP-17475:
  ssl: Do not run upgrade test when version is not changed
  ssl: Avoid unnecessary ASN1 decodes
  ssl: Extend certificate chain recreation logic
  public_key: Allow verify_fun to alter expire reason
  ssl: Remove dead code
  ssl: Use cert rec in db
  ssl: refactor some code
  ssl: Add a cert record containing both der and decoded
  ssl: avoid decoding certs several times

Commit:
Merge branch 'hasse/stdlib/chars_limit_fix/GH-5053/OTP-17525' into maint-24

* hasse/stdlib/chars_limit_fix/GH-5053/OTP-17525:
  stdlib: Fix a concerning io_lib option 'chars_limit'

# Conflicts:
#	lib/stdlib/test/io_SUITE.erl

Commit:
Merge branch 'ingela/ssl/ECDSA-signatures/OTP-17529' into maint-24

* ingela/ssl/ECDSA-signatures/OTP-17529:
  ssl: Fix signature algorithm selection for ECDSA

# Conflicts:
#	lib/ssl/test/ssl_cert_SUITE.erl

Commit:
Merge branch 'ingela/ssl/maint-24-pem-cache-test' into maint-24

* ingela/ssl/maint-24-pem-cache-test:
  ssl: Update test to use new parameter

Commit:
Merge branch 'ingela/ssl/runtime-dep' into maint-24

* ingela/ssl/runtime-dep:
  ssl: Update runtime dependency

# Conflicts:
#	lib/ssl/src/ssl.app.src

Commit:
Prepare release

Commit:
Updated OTP version

sigsergv committed:
Added tests and additional fix for public_key:pkix_verify

IngelaAndin committed:
Merge pull request #5064 from sigsergv/rsa_pss_sign_verify-test-fix

Fixing test public_key_SUITE:rsa_pss_sign_verify()

IngelaAndin committed:
Merge branch 'maint-23' into maint

* maint-23:
  Updated OTP version
  Prepare release

# Conflicts:
#	OTP_VERSION
#	erts/doc/src/notes.xml
#	erts/vsn.mk
#	lib/asn1/doc/src/notes.xml
#	lib/asn1/vsn.mk
#	lib/common_test/doc/src/notes.xml
#	lib/common_test/vsn.mk
#	lib/public_key/doc/src/notes.xml
#	lib/public_key/vsn.mk
#	lib/ssl/doc/src/notes.xml
#	lib/ssl/src/ssl.app.src
#	lib/ssl/vsn.mk
#	lib/stdlib/doc/src/notes.xml
#	lib/stdlib/vsn.mk
#	make/otp_version_tickets
#	otp_versions.table

IngelaAndin committed:
Merge branch 'maint-24' into maint

* maint-24:
  Updated OTP version
  Prepare release

# Conflicts:
#	lib/ssl/src/ssl.app.src
#	make/otp_version_tickets

sigsergv committed:
Fixed default pss_params for test suite

bmk committed:
[kernel|gen-udp-socket] Cosmetics

OTP-17410

IngelaAndin committed:
Merge pull request #5057 from sigsergv/pkix_sign_rsapss_fix-issue_5054

Fixing #5054, public_key:pkix_sign for RSASSA-PSS signature

OTP-17534

uabboli committed:
parsetools: Fix compiler warning in generated code

See also https://github.com/erlang/otp/issues/5067.

bmk committed:
[kernel|gen-tcp-socket] Tweak bind

It should be possible to bind to a port, but not address.
That is if you provide a port number but no address when
creating a socket (connect or listen). This case should be
mapped to #{port => Port, addr => any}.

OTP-17156

bmk committed:
[kernel|gen-tcp-socket] Handle buffered data when activating

If the socket is deactivated; active: once | true | N > 0 -> false
we do not cancel any select! Data that arrive during the phase when
we are in state 'recv' but are inactive is simply stored in the buffer.
Then if activated: active: false -> once | true | N > 0,
we need to check if there is something in our buffers, and maybe deliver
it to its owner. This should only occure if we are in state connected
(state 'recv' and in-active when data arrives => put data in buffer and
then enter state 'connected', since we are in-active).

OTP-17156

uabboli committed:
Merge branch 'hasse/parsetools/fix_yecc_bug/OTP-17535/GH-5067' into maint

* hasse/parsetools/fix_yecc_bug/OTP-17535/GH-5067:
  parsetools: Fix compiler warning in generated code

bmk committed:
[kernel|gen-tcp-socket] Extend handling of call-exceptions

Extend the number of 'exception's we explicitly "take care of"
in the call function.

OTP-17156

RaimoNiskanen committed:
Bind if port has been specified

RaimoNiskanen committed:
Correct guards for addr and port

RaimoNiskanen committed:
Test that setting only the source port for connect works

bmk committed:
Merge branch 'bmk/kernel/20210712/tweaking_kernel_tests' into maint

bmk committed:
Merge branch 'bmk/observer/20210713/update_app_depends/OTP-17346' into maint

OTP-17346

bmk committed:
Merge branch 'bmk/kernel/20210723/tweaking_bind_addr_and_activation/OTP-17156' into maint

OTP-17156

bmk committed:
Merge branch 'bmk/kernel/20210727/catch_shutdown_when_calling/OTP-17156' into maint

OTP-17156

RaimoNiskanen committed:
Merge branch 'raimo/kernel/tcp-bind-when-port-given/ERIERL-677/OTP-17536' into maint

* raimo/kernel/tcp-bind-when-port-given/ERIERL-677/OTP-17536:
  Correct guards for addr and port
  Bind if port has been specified
  Test that setting only the source port for connect works

Commit:
Merge branch 'raimo/kernel/tcp-bind-when-port-given/ERIERL-677/OTP-17536' into maint-24

* raimo/kernel/tcp-bind-when-port-given/ERIERL-677/OTP-17536:
  Correct guards for addr and port
  Bind if port has been specified
  Test that setting only the source port for connect works

Commit:
Prepare release

Commit:
Updated OTP version

RaimoNiskanen committed:
Merge branch 'maint-24' into maint

* maint-24:
  Updated OTP version
  Prepare release

# Conflicts:
#	make/otp_version_tickets

bmk committed:
Merge branch 'bmk/kernel/20210729/gen_udp_socket/OTP-17410' into maint

OTP-17410

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+
+  <p>Terminology used in this chapter:</p>
+  <list type="bulleted">
+    <item>A map with at most 32 elements will informally be called
```
... be called a <em>small map</em>.

bjorng replied:
Fixed.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+  <list type="bulleted">
+    <item>A map with at most 32 elements will informally be called
+    <em>small map</em>.</item>
+    <item>A map with more than 32 elements will informally be called
```
... be called a <em>large map</em>.

bjorng replied:
Fixed.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+      <item><p>When creating a new map, always create it with all keys
+      that will ever be used. To maximize sharing of keys (thus
+      minimizing memory use), create a single function that constructs
+      the map using the map syntax and always use it.</p></item>
```
> To maximize sharing of keys

Does this matter when the keys are atoms? If not, perhaps that is worth pointing out because atoms are probably common as keys when maps are used instead of records.

bjorng replied:
Yes, the size of the key tuple will be the same regardless of the type of the keys.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,618 @@
+
+      <item><p>Always update the map using the
+      <c>:=</c> operator (that is, requiring that an element with
+      that key already exists).  The <c>:=</c> operator is slightly
```
> ).  The <

Duplicated space (not sure if that matters)

bjorng replied:
Doesn't matter, but will fix anyway.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+
+      <item><p>Avoid default values and the <seemfa
+      marker="stdlib:maps#get/3">maps:get/3</seemfa>
+      function.</p></item>
```
Why does default values and maps:get/3 give bad performance?

kjellwinblad replied:
I think this item is connected with the next one. Perhaps it is better to merge them?

bjorng replied:
This is covered later. I said in the introduction that the those first sections would be brief, but that we would dig deeper in later sections.

I think that the items should be separate. The first item says to avoid default values, the other suggests **one** way to do so.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+
+    <p>Using a map as a dictionary implies the following usage pattern:</p>
+    <list type="bulleted">
+      <item><p>Keys are usually variables not known at compile-time.</p></item>
```
Keys are usually variables not known at compile-time.

Suggestions:

Keys are usually values not known at compile-time.

Keys are usually not known at compile-time.

Saying that keys are variables implies that they can change which I don't think is the case.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+
+    <p>Given that usage pattern, there is usually less performance to
+    be gained by using the map syntax compared to the <c>maps</c>
+    module. Therefore, which one to use is mostly a matter of
```
less performance to be gained compared to what?

Suggestion for replacement:

Given that usage pattern, the difference in performance between using the  map syntax and the <c>maps</c> module is usually small.

Out of curiosity, given the usage pattern is there any situation where the difference is not small? If there is no such situation, I think "usually" could be removed from the sentence.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+    taste.</p>
+
+    <p>Maps are usually the most efficient dictionary data structure,
+    with a few possible exceptions:</p>
```
I think "possible" can be removed.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+      marker="stdlib:sets#intersection/2">sets:intersection/2</seemfa>.
+      If the intersection operation is frequently used and operations
+      that operate on a single element in a set (such as
+      <c>is_element/2</c>) are avoided, <seeerl
```
gb_sets seems to have the same interface as sets as well and should at least in theory be able to have both efficient single item operations and a sets:intersection/2 that is much more efficient than the sets:intersection/2 for some input.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+        <item>The number of elements in the map.</item>
+
+        <tag>Keys</tag>
+        <item>A tuple with keys of the map: <c>{Key1,...,KeyN}</c>. The
```
A tuple with keys of the map: <c>{Key1,...,KeyN}</c>. The keys are sorted.</item> ->

A tuple with the keys of the map: <c>{Key1,...,KeyN}</c>. The keys are sorted in this tuple.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+
+      <p>When the value for an existing key is updated, the key tuple is not
+      be updated, allowing the key tuple to be shared with instances of the
+      map that have the same keys. In fact, the key tuple can be shared between all
```
be updated -> updated

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+      <p>When the value for an existing key is updated, the key tuple is not
+      be updated, allowing the key tuple to be shared with instances of the
+      map that have the same keys. In fact, the key tuple can be shared between all
+      instances of a map with some care. To arrange that, define a function that
```
instances of a map -> maps with the same keys

kjellwinblad replied:
So this is the motivation for some suggestions in the maps as record section. Perhaps, a reference from there to here would be useful?

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+      <p>Using the map syntax with small maps is particularly efficient.
+      As long as the keys are known at compile-time, the map is updated
+      in one go, making the time to update a map essentially constant regardless
+      of the number of keys updated. The same goes for matching.</p>
```
It is not clear to me why keys that are known at compile-time makes updates more efficient than when the keys are not known at compile time. Perhaps it can be good to explain this part.

bjorng replied:
It is explained later, but I will add an explanation to this paragraph too.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+      <p>When a large map is updated, the updated map and the original
+      map will share common parts of the HAMT, but sharing will never
+      be as effective as the sharing of the key tuple for a small
+      maps.</p>
```
Sharing of the keys of small maps is only efficient if the small maps use exactly the same keys and the "template" map is created in one go and in one place, if I understand it correctly. However, this sentence can be misunderstood as sharing of keys is always more efficient for small maps compared to large maps.

bjorng replied:
Will add "best possible" to the sentence:

> but sharing will never be as effective as the best possible sharing of the key tuple for small maps.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+      expected that many instances of the map will be created, it is
+      more efficient from a memory standpoint to avoid using large
+      maps (for example, by grouping related map elements into sub
+      maps to reduce the number of elements).</p>
```
One thing that might be good to mention is that the "template" map optimization is not the most memory efficient  representation if most "fields" has their default value. For example a map with 30 fields where only one field does not have the default value would use more memory with the "template" map optimization that without this optimization.

bjorng replied:
The template trick is all only meant to share the key tuple, not the array of values. The key tuple will be shared.

kjellwinblad replied:
Yes. You are right. I think my comment is misplaced. We recommend the template trick partly because it should reduce memory consumption. The point I was trying to make is that the template trick does not always reduce memory consumption so perhaps it can be useful to mention that.

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+    <code type="erl"><![CDATA[
+    Map2 = Map1#{Key1 := X},
+    Map3 = Map2#{Key2 := Y},
+    Map = Map3#{Key3 := Z}]]></code>
```
Not related to this document: I think this could be optimized. Instead of breaking up the update, one could check for duplicates among the keys and among duplicates of a particular key remove all that are not rightmost in the list. Is this something that is worth considering?

bjorng replied:
I don't think it's worth it. This is an unlikely usage for a small map, and for a large map the updates will be done sequentially anyway.

kjellwinblad replied:
Ok, perhaps a better approach would be to have a fast path that works most of the time but can fail over to what we are doing today if there are conflicts. But maybe it is not worth it anyway if it is not a so common to use non-literal keys?

kjellwinblad commented on system/doc/efficiency_guide/maps.xml:
```diff
@@ -0,0 +1,647 @@
+
+    <p>If a key is known to exist in a map, using the <c>:=</c>
+    operator is slightly more efficient than using the <c>=></c>
+    operator for small map.</p>
```
for small  map => for small maps

kjellwinblad reviewed (commented):
Review Part 1

I have not read the last section "Using the Functions in the maps Module" yet but I'm breaking it up in two parts so you can start to look at the comments for the other sections.

bjorng commented:
> Perhaps it can do more harm than good to give too specific advice. Users might start to do premature optimizations for efficiency reasons and create code that is hard to read.

Yes, that's always a risk. But the Efficiency Guide starts with a D.E. Knuth quote:

> Premature optimization is the root of all evil
